QModManager/API/Messages.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace QModManager.API
+{
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+    public class MessageReceiver : Attribute
+    {
+        public QMod Sender;
+
+        public MessageReceiver(QMod sender)
+        {
+            Sender = sender ?? throw new ArgumentNullException("The provided mod is null!", nameof(sender));
+        }
+        public MessageReceiver(Assembly senderAssembly)
+        {
+            Sender = QModAPI.GetMod(senderAssembly, true, true) ?? throw new ArgumentException("The provided assembly is not a mod assembly!", nameof(senderAssembly));
+        }
+        public MessageReceiver(string senderID)
+        {
+            // Should we not want to throw an exception in case the mod is not present?
+            Sender = QModAPI.GetMod(senderID, true, true) ?? throw new ArgumentException($"No mod matching the provided ID \"{senderID}\" was found!", nameof(senderID));
+        }
+    }
+
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+    public class GlobalMessageReciever : Attribute
+    {
+        public GlobalMessageReciever() { }
+    }
+
+    public static partial class QModAPI
+    {
+        public static void SendMessage(QMod mod, string message, params object[] data)
+        {
+            if (mod == null) return;
+            if (data == null) data = new object[] { };
+
+            QMod caller = GetMod(Assembly.GetCallingAssembly(), true);
+
+            if (mod.MessageReceivers.TryGetValue(caller, out List<MethodInfo> methods))
+                foreach (MethodInfo method in methods)
+                    method.Invoke(null, new object[] { caller, message, data });
+        }
+        public static void SendMessage(Assembly modAssembly, string message, params object[] data)
+            => SendMessage(GetMod(modAssembly, true, true), message, data);
+        public static void SendMessage(string modID, string message, params object[] data)
+            => SendMessage(GetMod(modID, true, true), message, data);
+
+        public static void BroadcastMessage(string message, params object[] data)
+        {
+            if (data == null) data = new object[] { };
+
+            QMod caller = GetMod(Assembly.GetCallingAssembly(), true);
+
+            if (GetAllMods().SelectMany(m => m.MessageReceivers)
+                            .ToDictionary(k => k.Key, v => v.Value)
+                            .TryGetValue(null, out List<MethodInfo> methods))
```

Comment:
If the methods are identified by `message` string, it looks like we could cache which mods implement that message.

When the mods are being loaded and the `MessageReceivers` are being cached, we could also build up data structure that quickly maps a method `message` identifier with the mods that implement that message:  
```CSharp
internal readonly static Dictionary<string, List<QMod>> CachedMessageReceivers = new Dictionary<string, List<QMod>>();
```
This would allow for less searching in the `BroadcastMessage` and `SendMessage` methods.
```CSharp
if (CachedMessageReceivers.TryGetValue(message, out List<QMod> mods))
{
    foreach (QMod mod in mods)
        foreach (MethodInfo method in mod.MessageReceivers)
            method.Invoke(null, new object[] { caller, message, data });
}
```
Making sure this feature doesn't introduce performance penalties will encourage modders to use it.

Reply:
Methods are not identified by the `message` string, it is used to identify what kind of message is sent and how the `data` object will be used. It might as well have been a part of the `data` array.